Add support for other workers/worklets

hughrawlinson opened the pull request:
Sorry for the delay - here's the beginnings of a PR to address #52.

It works just fine in my own test. But right now it only supports Base64 Audio Worklets. The other things that need support are:

- Base64
  - [x] Audio Worklet
  - [x] Paint Worklet
  - [x] Service Worker
  - [x] Shared Worker
  - ~Animation Worklet~
  - ~Layout Worklet~
- Inline
  - [x] Audio Worklet
  - [x] Paint Worklet
  - [x] Service Worker
  - [x] Shared Worker
  - ~Animation Worklet~
  - ~Layout Worklet~
- URL
  - [x] Audio Worklet
  - [x] Paint Worklet
  - [x] Service Worker
  - [x] Shared Worker
  - ~Animation Worklet~
  - ~Layout Worklet~
- [x] Documentation

Given how much there is to add support for, I wonder if it's worth changing the structure of the helper directory a little so that there isn't so much riding on the filenames?

A problem that cropped up is that AudioWorklet doesn't exist in Node. I have the helper method throwing an error, but would love to hear your take on what should be done about that. Would it be better to fail during compile time based on the target?

It's also not clear that Paint Worklet and Animation Worklet are widely supported. I'll speak to some people and try to figure out the utility at this point. Once AudioWorklet is in, adding the others should be fairly straightforward based on that example. Web Audio API finally made it to Recommendation status at the W3C last week.

Good news is that I don't think this patch will require a major version bump - I was able to keep the original 'pattern' option around, so that this isn't a breaking change.

I plan to continue working on adding support for the other Worklets in the matrix above - although depending on what you think it might make sense to split this work out into multiple PRs rather than have all in one - and maybe first focus on AudioWorklet?

Also, is there a testing strategy? I would love to provide some tests for this if that's useful - just not really sure how to.

Thanks!

code2933 committed:
feat: SharedWorker support

Co-Authored-By: Hugh Rawlinson <hughr2005@gmail.com>

hughrawlinson committed:
working mvp of base64audioworkletfactory

hughrawlinson committed:
add support for urlaudioworklets

hughrawlinson committed:
Add support for Audio Worklets preserveSource

- additionally, rename a function for easier debugging

darionco commented on src/index.js:
```diff
@@ -26,6 +30,9 @@ const defaultConfig = {
 };
 
 module.exports = function workerLoaderPlugin(userConfig = null) {
+    if (userConfig) {
+        userConfig.webWorkerPattern = userConfig?.webWorkerPattern || userConfig?.pattern || null;
+    }
```
I think we could release this as a major version so breaking backwards compatibility would be ok. I am thinking that having `pattern` and `*WorkerPattern` could get confusing for new users.

hughrawlinson replied:
Ok, I'll remove these few lines then 😄

darionco commented:
Nice! I am super excited about this PR!

hughrawlinson commented:
Awesome, great to hear! Any thoughts as to the above questions? Also I noticed that #49 is open and adds another worker that I had missed on my list. Shall we add that to the major release?

hughrawlinson committed:
remove backward compat for pattern option

hughrawlinson committed:
add support for paint worklets

hughrawlinson committed:
make paintworklet not async

hughrawlinson committed:
remove scaffolding for animationworklet

animationworklet and layout worklet don't have implementations in any
browsers yet so I'll suggest that we add them as a feature once that
happens.

hughrawlinson commented on src/helper/browser/createURLPaintWorkletFactory.js:
```diff
@@ -0,0 +1,6 @@
+export function createURLPaintWorkletFactory(url) {
+    return async function PaintWorkletFactory(options) {
+        return await CSS.paintWorklet.addModule(url, options);
```
So there's a minor issue here. `paintWorklet` isn't necessarily supported in all browsers, and as such, `CSS.paintWorklet` may not be defined. We have a few options.

- Leave code as-is to throw the referenceerror. Pros: we don't have to take an error handling stance. Cons: it might not be clear to the developer what's happening.
- Catch the error and throw a more explanatory error
- In the case that it fails, return some other object from the function. Pros: we won't throw runtime exceptions. Cons: it might be unclear to the developer when their paint call in CSS just doesn't work.
- Log a debug statement rather than throw an error?

Let me know what approach you'd like me to take here.

hughrawlinson committed:
add support for service workers

hughrawlinson committed:
add docs

hughrawlinson commented:
Ok I think we're close to there, certainly ready for a review. The outstanding questions are:

- There aren't any tests. I've manually tested as much as I can, but would it be worth getting some automated tests in place?
- I pulled in @code2933's code for sharedworkers from #49 since they're another worker type here. LMK if that isn't ok, happy to remove them from this PR.
- What should we do about throwing errors in node when the worker/worklet type isn't available? And in browsers that don't support for example paintWorklet (IE).
- I removed AnimationWorklet and LayoutWorklet from the scope of this PR since there are no implementations in any browsers. It should be fairly straightforward to look at this PR and add support for them once/if they do make it to browsers.
- Are the docs ok? Anything more needs clarifying?
- Almost definitely a separate PR, but one that I'm happy to do and might fit in with releasing 2.0.0 - how to you feel about adding typescript types? The declaration file in the repo you link to is good, and I could add additional types for the other worklet types. Or, if you want to keep this repo typescript free, we could put types up on DefinitelyTyped.

Let me know if anything else is unclear, of course happy to answer any questions or make any changes.

hughrawlinson committed:
add bundle-time error for non-web-workers where target is node

darionco commented:
This is so awesome, I'm so sorry it took this long to get to it, life stuff and all that.

Thank you so much for all this work!

darionco commented:
Let me know if you'd be interested on maintaining this project, unfortunately, my current situation will not allow me to continue with development. :(

hughrawlinson commented:
Thanks @darionco! I would certainly consider that 😄 

Would you be able to publish a new version of the package with this change?

darionco commented:
@hughrawlinson I have invited you as a collaborator for this project, if you give me your npm username I'll add you as collaborator there as well so you can publish new versions of the package.

hughrawlinson commented:
Thanks! Same username as here - https://www.npmjs.com/~hughrawlinson 😄

darionco commented:
Done, thank you for your contributions!

Let me know if I can help in any other way!

jaimeadf commented:
Hey, will these changes get a new version in NPM? I have the latest version `1.6.1` installed, and it only supports web workers.

![image](https://user-images.githubusercontent.com/40345645/169674236-e0847e04-941c-4d26-ad5c-9f14edc5897d.png)
